lib/async_hooks.js

```diff
@@ -98,24 +103,19 @@ class AsyncHook {
     if (hooks_array.includes(this))
       return;
 
-    if (!setupHooksCalled) {
-      setupHooksCalled = true;
-      // Setup the callbacks that node::AsyncWrap will call when there are
-      // hooks to process. They use the same functions as the JS embedder API.
-      async_wrap.setupHooks({ init,
-                              before: emitBeforeN,
-                              after: emitAfterN,
-                              destroy: emitDestroyN });
-    }
-
     // createHook() has already enforced that the callbacks are all functions,
     // so here simply increment the count of whether each callbacks exists or
     // not.
     hook_fields[kInit] += +!!this[init_symbol];
     hook_fields[kBefore] += +!!this[before_symbol];
     hook_fields[kAfter] += +!!this[after_symbol];
     hook_fields[kDestroy] += +!!this[destroy_symbol];
+    hook_fields[kTotals]++;
```

Comment:
I asked @matthewloring about this on IRC:
```
<trevnorris> if an async hook is enabled that has no callbacks should we go through
             the trouble of setting up the PromiseWrap in async_wrap.cc PromiseHook?
<mattloring> Is it expected to be a common case to have async hooks enabled with no
             listeners? If listeners are added later then it might be good to have
             PromiseWraps available to resolve parent promise ids
<trevnorris> it isn't a common case, but i can see your point. so continue to add
             PromiseWraps to Promises even if there are no callbacks to be called.
```
@AndreasMadsen My original PR tracked the number of callbacks to be called, and only called `enablePromiseHook()` if the number of callbacks was `> 0`. If you think it should be difference feel free to chime in.